// @spore-sdk/core/lib/api/composed/spore/meltSpore.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.meltSpore = void 0;
const lumos_1 = require("@ckb-lumos/lumos");
const helpers_1 = require("../../../helpers");
const config_1 = require("../../../config");
const __1 = require("../..");
async function meltSpore(props) {
    // Env
    const config = props.config ?? (0, config_1.getSporeConfig)();
    const indexer = new lumos_1.Indexer(config.ckbIndexerUrl, config.ckbNodeUrl);
    // Get TransactionSkeleton
    let txSkeleton = lumos_1.helpers.TransactionSkeleton({
        cellProvider: indexer,
    });
    // Inject live spore to Transaction.inputs
    const sporeCell = await (0, __1.getSporeByOutPoint)(props.outPoint, config);
    const injectLiveSporeCellResult = await (0, __1.injectLiveSporeCell)({
        cell: sporeCell,
        updateWitness: props.updateWitness,
        txSkeleton,
        config,
    });
    txSkeleton = injectLiveSporeCellResult.txSkeleton;
    // Inject needed capacity and pay fee
    const injectCapacityAndPayFeeResult = await (0, helpers_1.injectCapacityAndPayFee)({
        txSkeleton,
        changeAddress: props.changeAddress,
        fromInfos: props.fromInfos,
        fee: lumos_1.BI.from(0),
        config,
    });
    txSkeleton = injectCapacityAndPayFeeResult.txSkeleton;
    return {
        txSkeleton,
        inputIndex: injectLiveSporeCellResult.inputIndex,
    };
}
exports.meltSpore = meltSpore;
