// @spore-sdk/core/lib/api/composed/cluster/transferCluster.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transferCluster = void 0;
const lumos_1 = require("@ckb-lumos/lumos");
const config_1 = require("../../../config");
const helpers_1 = require("../../../helpers");
const __1 = require("../..");
async function transferCluster(props) {
    // Env
    const config = props.config ?? (0, config_1.getSporeConfig)();
    const indexer = new lumos_1.Indexer(config.ckbIndexerUrl, config.ckbNodeUrl);
    const useCapacityMarginAsFee = props.useCapacityMarginAsFee ?? true;
    // Check capacity margin related props
    if (!useCapacityMarginAsFee && !props.fromInfos?.length) {
        throw new Error('When useCapacityMarginAsFee is enabled, fromInfos is also required');
    }
    if (useCapacityMarginAsFee && props.capacityMargin !== void 0) {
        throw new Error('When useCapacityMarginAsFee is enabled, cannot set capacityMargin of the cluster');
    }
    // Get TransactionSkeleton
    let txSkeleton = lumos_1.helpers.TransactionSkeleton({
        cellProvider: indexer,
    });
    // Find cluster by OutPoint
    const clusterCell = await (0, __1.getClusterByOutPoint)(props.outPoint, config);
    // Add cluster to Transaction.inputs and Transaction.outputs
    const injectLiveClusterCellResult = await (0, __1.injectLiveClusterCell)({
        txSkeleton,
        addOutput: true,
        cell: clusterCell,
        updateWitness: props.updateWitness,
        capacityMargin: props.capacityMargin,
        updateOutput(cell) {
            cell.cellOutput.lock = props.toLock;
            if (props.updateOutput instanceof Function) {
                cell = props.updateOutput(cell);
            }
            return cell;
        },
        config,
    });
    txSkeleton = injectLiveClusterCellResult.txSkeleton;
    if (!useCapacityMarginAsFee) {
        // Inject needed capacity and pay fee
        const injectCapacityAndPayFeeResult = await (0, helpers_1.injectCapacityAndPayFee)({
            txSkeleton,
            changeAddress: props.changeAddress,
            fromInfos: props.fromInfos,
            fee: lumos_1.BI.from(0),
            config,
        });
        txSkeleton = injectCapacityAndPayFeeResult.txSkeleton;
    }
    else {
        // Pay fee by the cluster cell's capacity margin
        txSkeleton = await (0, helpers_1.payFeeByOutput)({
            outputIndex: injectLiveClusterCellResult.outputIndex,
            txSkeleton,
            config,
        });
    }
    return {
        txSkeleton,
        inputIndex: injectLiveClusterCellResult.inputIndex,
        outputIndex: injectLiveClusterCellResult.outputIndex,
    };
}
exports.transferCluster = transferCluster;
